src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala

```diff
@@ -39,40 +40,55 @@ private[tracker] object TaskOpProcessorImpl {
       implicit ec: ExecutionContext): Future[Action] = {
       directTaskTracker.task(taskId).map {
         case Some(existingTask) =>
-          resolveForExistingTask(existingTask, status)
+          actionForTaskAndStatus(existingTask, status)
         case None =>
           Action.Fail(new IllegalStateException(s"$taskId of app [${taskId.appId}] does not exist"))
       }
     }
 
+    private[this] def actionForTaskAndStatus(task: Task, statusUpdate: TaskStatus): Action = {
+      //Step 1: if a resident task is terminated, it has to be handled different
+      resolveForResidentFailed(task, statusUpdate)
+        //Step 2: update existing running task
+        .orElse(resolveForLaunchedTask(task, statusUpdate))
+        //Step 3: nothing of the above: do nothing
+        .getOrElse {
+          log.warn("Got update for task which wasn't launched yet: {}", statusUpdate)
+          Action.Noop
+        }
+    }
+    private[this] def resolveForResidentFailed(task: Task, statusUpdate: TaskStatus): Option[Action] = {
+      for {
+        reservation <- task.reservationWithVolumes
+        launched <- task.launched
+        if Terminated.isTerminated(statusUpdate.getState)
+      } yield Unlaunch(task)
+    }
+
     /**
       * Calculates the change that needs to performed on this task according to the given task status update
       */
-    private[this] def resolveForExistingTask(taskState: Task, statusUpdate: TaskStatus): Action = {
-      taskState.launched match {
-        case Some(currentLaunched) =>
-          statusUpdate.getState match {
-            case TASK_ERROR | TASK_FAILED | TASK_FINISHED | TASK_KILLED | TASK_LOST =>
-              Action.Expunge
-            case TASK_RUNNING if !currentLaunched.hasStartedRunning => // was staged, is now running
-              val now = clock.now()
-              Action.Update(
-                taskState.copy(launched = Some(
-                  currentLaunched.copy(
-                    status = currentLaunched.status.copy(
-                      startedAt = Some(now),
-                      mesosStatus = Some(statusUpdate)
-                    )
-                  )
-                ))
+    private[this] def resolveForLaunchedTask(task: Task, statusUpdate: TaskStatus): Option[Action] = {
+      def stagedNowRunning(currentLaunched: Task.Launched): Action = {
```

Comment:
`stagedNowRunning` is defined in local scope of `resolveForLaunchedTask`, `updateTaskOnStateChange` is defined outside of this scope. Both are only used from within `resolveForLaunchedTask`, so I'd suggest to define both in the same scope (I don't really mind which one).